From 25eb5856da48dd885125b7e193b41277b0d73e09 Mon Sep 17 00:00:00 2001
From: Shahed Shaikh <shahedx.shaikh@intel.com>
Date: Thu, 16 Jul 2020 18:13:05 +0800
Subject: [PATCH] DRV_DPM_SW-329: dp_xmit minor enhancement

- print tx flags strings instead of hex value
- use common function dp_tx_dbg() instead of HAL level debug

Signed-off-by: Shahed Shaikh <shahedx.shaikh@intel.com>
---
 drivers/net/datapath/dpm/datapath_tx.c         | 35 +++++++++-
 drivers/net/datapath/dpm/gswip30/datapath_tx.c | 92 ++------------------------
 drivers/net/datapath/dpm/gswip31/datapath_tx.c | 92 ++------------------------
 3 files changed, 41 insertions(+), 178 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath_tx.c b/drivers/net/datapath/dpm/datapath_tx.c
index 475d96f8044a..38b35795b4b2 100644
--- a/drivers/net/datapath/dpm/datapath_tx.c
+++ b/drivers/net/datapath/dpm/datapath_tx.c
@@ -5,6 +5,7 @@
 
 #include <linux/kernel.h>
 #include <linux/types.h>
+#include <linux/log2.h>
 #include "datapath.h"
 #include "datapath_tx.h"
 
@@ -54,6 +55,20 @@ struct dp_tx_context {
 
 static struct dp_tx_context *dp_tx_ctx;
 
+#define POS(VAL) ilog2(DP_TX_##VAL)
+const char *dp_tx_flag_str[] = {
+	[0] 				= "NONE",
+	[POS(DSL_FCS)] 		= "DP_TX_DSL_FCS",
+	[POS(INSERT_PMAC)] 	= "DP_TX_INSERT_PMAC",
+	[POS(OAM)] 			= "DP_TX_OAM",
+	[POS(TO_DL_MPEFW)] 	= "DP_TX_TO_DL_MPEFW",
+	[POS(INSERT)] 		= "DP_TX_INSERT",
+	[POS(INSERT_POINT)] = "DP_TX_INSERT_POINT",
+};
+
+#define dp_tx_flags_str_len \
+		(sizeof (dp_tx_flag_str) / sizeof (dp_tx_flag_str[0]))
+
 void dp_tx_dbg(char *title, struct sk_buff *skb, s32 ep, s32 len, u32 flags,
 	       struct pmac_tx_hdr *pmac, struct dp_subif *rx_subif,
 	       int need_pmac, int gso, int checksum)
@@ -113,8 +128,24 @@ void dp_tx_dbg(char *title, struct sk_buff *skb, s32 ep, s32 len, u32 flags,
 	;
 #endif
 
-	DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "flags=0x%x skb->len=%d skb->priority=%d\n",
-		 flags, skb->len, skb->priority);
+	DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "skb->len=%d skb->priority=%d\n",
+		 skb->len, skb->priority);
+
+	if (dp_dbg_flag & DP_DBG_FLAG_DUMP_TX) {
+		int i;
+		pr_info("flags =");
+		if (flags) {
+			/* valid flags start from index 1 */
+			for (i = 1; i < dp_tx_flags_str_len; i++) {
+				if (flags & (1 << i))
+					pr_info(" %s |", dp_tx_flag_str[i]);
+			}
+		} else {
+			pr_info(" %s", dp_tx_flag_str[0]);
+		}
+		pr_info("\n");
+	}
+
 	DP_DEBUG(DP_DBG_FLAG_DUMP_TX,
 		 "skb->data=0x%px with pmac hdr size=%zu\n", skb->data,
 		 sizeof(struct pmac_tx_hdr));
diff --git a/drivers/net/datapath/dpm/gswip30/datapath_tx.c b/drivers/net/datapath/dpm/gswip30/datapath_tx.c
index 26ee7e7b11cf..c6459719d2c0 100644
--- a/drivers/net/datapath/dpm/gswip30/datapath_tx.c
+++ b/drivers/net/datapath/dpm/gswip30/datapath_tx.c
@@ -14,91 +14,7 @@
 #include "../datapath.h"
 #include "datapath_misc.h"
 #include "../datapath_instance.h"
-
-void dp_xmit_dbg(
-	char *title,
-	struct sk_buff *skb,
-	s32 ep,
-	s32 len,
-	u32 flags,
-	struct pmac_tx_hdr *pmac,
-	dp_subif_t *rx_subif,
-	int need_pmac,
-	int gso,
-	int checksum)
-{
-	int data_len;
-
-#ifdef DP_SKB_HACK
-	DP_DEBUG(DP_DBG_FLAG_DUMP_TX,
-		 "%s: dp_xmit:skb->data/len=0x%p/%d data_ptr=%x from port=%d and subitf=%d\n",
-		 title,	skb->data, len,
-		 ((struct dma_tx_desc_2 *)&skb->DW2)->field.data_ptr,
-		 ep, rx_subif->subif);
-#endif
-	if (dp_dbg_flag & DP_DBG_FLAG_DUMP_TX_DATA) {
-		data_len = skb->len > dp_print_len ? skb->len : dp_print_len;
-		if (pmac)
-			dp_dump_raw_data((char *)pmac, PMAC_SIZE, "Tx Data");
-		dp_dump_raw_data(skb->data, data_len, "Tx Data");
-	}
-	DP_DEBUG(DP_DBG_FLAG_DUMP_TX_SUM,
-		 "ip_summed=%s(%d) encapsulation=%s\n",
-		 dp_skb_csum_str(skb), skb->ip_summed,
-		 skb->encapsulation ? "Yes" : "No");
-	if (skb->encapsulation)
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX_SUM,
-			 "inner ip start=0x%x(%d), transport=0x%x(%d)\n",
-			 (unsigned int)skb_inner_network_header(skb),
-			 (int)(skb_inner_network_header(skb) -
-			 skb->data),
-			 (unsigned int)
-			 skb_inner_transport_header(skb),
-			 (int)(skb_inner_transport_header(skb) -
-			 skb_inner_network_header(skb)));
-	else
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX_SUM,
-			 "ip start=0x%x(%d), transport=0x%x(%d)\n",
-			 (unsigned int)(unsigned int)
-			 skb_network_header(skb),
-			 (int)(skb_network_header(skb) - skb->data),
-			 (unsigned int)skb_transport_header(skb),
-			 (int)(skb_transport_header(skb) -
-			 skb_network_header(skb)));
-
-	if (dp_dbg_flag & DP_DBG_FLAG_DUMP_TX_DESCRIPTOR)
-#ifdef DP_SKB_HACK
-		dp_port_prop[0].info.dump_tx_dma_desc(
-				(struct dma_tx_desc_0 *)&skb->DW0,
-				(struct dma_tx_desc_1 *)&skb->DW1,
-				(struct dma_tx_desc_2 *)&skb->DW2,
-				(struct dma_tx_desc_3 *)&skb->DW3);
-#endif
-
-	DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "flags=0x%x skb->len=%d\n",
-		 flags, skb->len);
-	DP_DEBUG(DP_DBG_FLAG_DUMP_TX,
-		 "skb->data=0x%p with pmac hdr size=%u\n", skb->data,
-		 sizeof(struct pmac_tx_hdr));
-	if (need_pmac) { /*insert one pmac header */
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX,
-			 "need pmac\n");
-		if (pmac && (dp_dbg_flag & DP_DBG_FLAG_DUMP_TX_DESCRIPTOR))
-			dp_port_prop[0].info.dump_tx_pmac(pmac);
-	} else {
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "no pmac\n");
-	}
-	if (gso)
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "GSO pkt\n");
-	else
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "Non-GSO pkt\n");
-	if (checksum)
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "Need checksum offload\n");
-	else
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "No need checksum offload pkt\n");
-
-	DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "\n\n");
-}
+#include "../datapath_tx.h"
 
 int32_t dp_xmit_30(struct net_device *rx_if, dp_subif_t *rx_subif,
 		   struct sk_buff *skb, int32_t len, uint32_t flags)
@@ -161,7 +77,7 @@ int32_t dp_xmit_30(struct net_device *rx_if, dp_subif_t *rx_subif,
 		ltq_mpe_fasthook_tx_fn(skb, 0, NULL);
 #endif
 	if (unlikely(dp_dbg_flag))
-		dp_xmit_dbg("\nOrig", skb, ep, len, flags,
+		dp_tx_dbg("\nOrig", skb, ep, len, flags,
 			    NULL, rx_subif, 0, 0, flags & DP_TX_CAL_CHKSUM);
 
 	/*No PMAC for WAVE500 and DSL by default except bonding case */
@@ -346,11 +262,11 @@ int32_t dp_xmit_30(struct net_device *rx_if, dp_subif_t *rx_subif,
 
 	if (unlikely(dp_dbg_flag)) {
 		if (insert_pmac_f)
-			dp_xmit_dbg("After", skb, ep, len, flags, &pmac,
+			dp_tx_dbg("After", skb, ep, len, flags, &pmac,
 				    rx_subif, insert_pmac_f, skb_is_gso(skb),
 				    tx_chksum_flag);
 		else
-			dp_xmit_dbg("After", skb, ep, len, flags, NULL,
+			dp_tx_dbg("After", skb, ep, len, flags, NULL,
 				    rx_subif, insert_pmac_f, skb_is_gso(skb),
 				    tx_chksum_flag);
 	}
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_tx.c b/drivers/net/datapath/dpm/gswip31/datapath_tx.c
index 83332d09ad1d..7c684bd626f5 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_tx.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_tx.c
@@ -10,91 +10,7 @@
 #include "../datapath.h"
 #include "datapath_misc.h"
 #include "../datapath_instance.h"
-
-void dp_xmit_dbg(
-	char *title,
-	struct sk_buff *skb,
-	s32 ep,
-	s32 len,
-	u32 flags,
-	struct pmac_tx_hdr *pmac,
-	dp_subif_t *rx_subif,
-	int need_pmac,
-	int gso,
-	int checksum)
-{
-	int data_len;
-
-#if defined(DP_SKB_HACK)
-	DP_DEBUG(DP_DBG_FLAG_DUMP_TX,
-		 "%s: dp_xmit:skb->data/len=0x%p/%d data_ptr=%x from port=%d and subitf=%d\n",
-		 title,	skb->data, len,
-		 ((struct dma_tx_desc_2 *)&skb->DW2)->field.data_ptr,
-		 ep, rx_subif->subif);
-#endif
-	if (dp_dbg_flag & DP_DBG_FLAG_DUMP_TX_DATA) {
-		data_len = skb->len > dp_print_len ? skb->len : dp_print_len;
-		if (pmac)
-			dp_dump_raw_data((char *)pmac, PMAC_SIZE, "Tx Data");
-		dp_dump_raw_data(skb->data, data_len, "Tx Data");
-	}
-	DP_DEBUG(DP_DBG_FLAG_DUMP_TX_SUM,
-		 "ip_summed=%s(%d) encapsulation=%s\n",
-		 dp_skb_csum_str(skb), skb->ip_summed,
-		 skb->encapsulation ? "Yes" : "No");
-	if (skb->encapsulation)
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX_SUM,
-			 "inner ip start=0x%x(%d), transport=0x%x(%d)\n",
-			 (unsigned int)skb_inner_network_header(skb),
-			 (int)(skb_inner_network_header(skb) -
-			       skb->data),
-			 (unsigned int)
-			 skb_inner_transport_header(skb),
-			 (int)(skb_inner_transport_header(skb) -
-			       skb_inner_network_header(skb)));
-	else
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX_SUM,
-			 "ip start=0x%x(%d), transport=0x%x(%d)\n",
-			 (unsigned int)(unsigned int)
-			 skb_network_header(skb),
-			 (int)(skb_network_header(skb) - skb->data),
-			 (unsigned int)skb_transport_header(skb),
-			 (int)(skb_transport_header(skb) -
-			       skb_network_header(skb)));
-
-	if (dp_dbg_flag & DP_DBG_FLAG_DUMP_TX_DESCRIPTOR)
-#if defined(DP_SKB_HACK)
-		dp_port_prop[0].info.dump_tx_dma_desc(
-			(struct dma_tx_desc_0 *)&skb->DW0,
-			(struct dma_tx_desc_1 *)&skb->DW1,
-			(struct dma_tx_desc_2 *)&skb->DW2,
-			(struct dma_tx_desc_3 *)&skb->DW3);
-#endif
-
-	DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "flags=0x%x skb->len=%d\n",
-		 flags, skb->len);
-	DP_DEBUG(DP_DBG_FLAG_DUMP_TX,
-		 "skb->data=0x%p with pmac hdr size=%u\n", skb->data,
-		 sizeof(struct pmac_tx_hdr));
-	if (need_pmac) { /*insert one pmac header */
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX,
-			 "need pmac\n");
-		if (pmac && (dp_dbg_flag & DP_DBG_FLAG_DUMP_TX_DESCRIPTOR))
-			dp_port_prop[0].info.dump_tx_pmac(pmac);
-	} else {
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "no pmac\n");
-	}
-	if (gso)
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "GSO pkt\n");
-	else
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "Non-GSO pkt\n");
-	if (checksum)
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "Need checksum offload\n");
-	else
-		DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "No need checksum offload pkt\n");
-
-	DP_DEBUG(DP_DBG_FLAG_DUMP_TX, "\n\n");
-}
+#include "../datapath_tx.h"
 
 static int dp_reinsert_pkt_cnt(struct pmac_port_info *dp_info, int vap,
 			       struct sk_buff *skb)
@@ -187,7 +103,7 @@ int32_t dp_xmit_31(struct net_device *rx_if, dp_subif_t *rx_subif,
 		ltq_mpe_fasthook_tx_fn(skb, 0, NULL);
 #endif
 	if (unlikely(dp_dbg_flag))
-		dp_xmit_dbg("\nOrig", skb, ep, len, flags,
+		dp_tx_dbg("\nOrig", skb, ep, len, flags,
 			    NULL, rx_subif, 0, 0, flags & DP_TX_CAL_CHKSUM);
 
 	/*No PMAC for WAVE500 and DSL by default except bonding case */
@@ -372,11 +288,11 @@ int32_t dp_xmit_31(struct net_device *rx_if, dp_subif_t *rx_subif,
 
 	if (unlikely(dp_dbg_flag)) {
 		if (insert_pmac_f)
-			dp_xmit_dbg("After", skb, ep, len, flags, &pmac,
+			dp_tx_dbg("After", skb, ep, len, flags, &pmac,
 				    rx_subif, insert_pmac_f, skb_is_gso(skb),
 				    tx_chksum_flag);
 		else
-			dp_xmit_dbg("After", skb, ep, len, flags, NULL,
+			dp_tx_dbg("After", skb, ep, len, flags, NULL,
 				    rx_subif, insert_pmac_f, skb_is_gso(skb),
 				    tx_chksum_flag);
 	}
